pkg/manager/member/pvc_resizer.go

```diff
@@ -385,132 +375,290 @@ func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
 
 	// sync volume status for `sourceVolumeStatus`
 	for volName, status := range observedStatus {
-		if _, exist := ctx.sourceVolumeStatus[volName]; !exist {
-			ctx.sourceVolumeStatus[volName] = &v1alpha1.StorageVolumeStatus{
+		if _, exist := sourceVolumeStatus[volName]; !exist {
+			sourceVolumeStatus[volName] = &v1alpha1.StorageVolumeStatus{
 				Name: volName,
 			}
 		}
-		ctx.sourceVolumeStatus[volName].ObservedStorageVolumeStatus = *status
+		sourceVolumeStatus[volName].ObservedStorageVolumeStatus = *status
 	}
-	for _, status := range ctx.sourceVolumeStatus {
+	for _, status := range sourceVolumeStatus {
 		if _, exist := observedStatus[status.Name]; !exist {
-			delete(ctx.sourceVolumeStatus, status.Name)
+			delete(sourceVolumeStatus, status.Name)
 		}
 	}
 }
 
 // resizeVolumes resize PVCs by comparing `desiredVolumeQuantity` and `actualVolumeQuantity` in context.
 func (p *pvcResizer) resizeVolumes(ctx *componentVolumeContext) error {
-	desiredVolumeQuantity := ctx.desiredVolumeQuantity
-	podVolumes := ctx.actualPodVolumes
+	var (
+		resizingPod       *corev1.Pod
+		classifiedVolumes map[volumePhase][]*volume
+	)
+
+	// choose one pod whose volume need to be resized
+	for _, podVolumes := range ctx.actualPodVolumes {
+		curClassifiedVolumes, err := p.classifyVolumes(ctx, podVolumes.volumes)
+		if err != nil {
+			return fmt.Errorf("classify volumes for %s failed: %w", ctx.ComponentID(), err)
+		}
 
-	id := fmt.Sprintf("%s/%s", ctx.namespace, ctx.name)
+		if len(curClassifiedVolumes[resizing]) != 0 || len(curClassifiedVolumes[needResize]) != 0 {
+			resizingPod = podVolumes.pod
+			classifiedVolumes = curClassifiedVolumes
+			break
+		}
+	}
 
-	for _, podVolume := range podVolumes {
+	allResized := resizingPod == nil
+	condExist := meta.IsStatusConditionTrue(ctx.status.GetConditions(), v1alpha1.ComponentVolumeResizing)
 
-		allResized := true
-		errors := []error{}
+	if allResized {
+		if condExist {
+			return p.endResize(ctx)
+		}
+		klog.V(4).Infof("all volumes are resized for %s", ctx.ComponentID())
+	} else {
+		if !condExist {
+			return p.beginResize(ctx)
+		}
 
-		for volName, pvc := range podVolume.volToPVCs {
-			pvcID := fmt.Sprintf("%s/%s", pvc.Namespace, pvc.Name)
+		// some volumes need to be resized
+		if len(classifiedVolumes[needResize]) != 0 {
+			klog.V(4).Infof("start to resize volumes of Pod %s/%s for %s", resizingPod.Namespace, resizingPod.Name, ctx.ComponentID())
+			return p.resizeVolumesForPod(ctx, resizingPod, classifiedVolumes[needResize])
```

Comment:
This is pod-specific logic, can we separate it from the condition update logic which is component-level logic?